### PR TITLE
[Merged] 1.add gesture event;2.fixed draw arc round bug;

### PR DESCRIPTION
**Pull request by Simen-z**
Recently I have been busy porting littlevgl to stm32l4 and a few other things. Sorry for not coming and submitting the code.
1.add gesture event and use the sample:

```
...
lv_obj_t *obj = lv_obj_create(lv_disp_get_scr_act(NULL), NULL);
lv_obj_set_gusture(obj, true);
lv_obj_set_event_cb(obj , event_cb);
...
void event_cb(lv_obj_t * obj, lv_event_t event)
{
	lv_indev_t * pIndev = lv_indev_get_act();
	if (pIndev == NULL)
		return;

	switch (event){
	case LV_EVENT_GESTURE:{
                      lv_gesture_dir_t gesture_dir = lv_indev_get_gesture_dir(pIndev);
                      switch(gesture_dir ){
                      case 	LV_GESTURE_DIR_TOP:		/**< Gesture dir up. */
                          /*on led 1*/
                      break;
                      case 	LV_GESTURE_DIR_BOTTOM:	/**< Gesture dir down. */
                          /*on led 2*/
                      break;
                      case 	LV_GESTURE_DIR_LEFT:	/**< Gesture dir left. */
                          /*on led 3*/
                      break;
                      case 	LV_GESTURE_DIR_RIGHT:	/**< Gesture dir right. */
                          /*on led 4*/
                      break;
                     }
                }
		break;
	}
}
```
2.fixed draw arc round bug:in function
`static void get_rounded_area(int16_t angle, lv_coord_t radius, uint8_t tickness, lv_area_t * res_area);`
```
    lv_coord_tcir_x;
    lv_coord_tcir_y;

    cir_x = ((radius - rx_corr - thick_half) * lv_trigo_sin(90 - angle)) >> (LV_TRIGO_SHIFT - ps);
    cir_y = ((radius - ry_corr - thick_half) * lv_trigo_sin(angle)) >> (LV_TRIGO_SHIFT - ps);
```
Data of type lv_coord_t will overflow during calculation and be modified to int32_t;

### CODE REVIEW

**Comment by kisvegabor** [comment]
Thank you!

I've read it through and it looks good! 
I'm thinking about enabling `gesture_parent` by default. This way the gestures are propagated to the screen and can be easily processed there. I think the most probable scenario is to use screen-sized gestures.

What do you think?


---
Another idea: what about clearing `gesture_in_prog` when `drag_in_prog` is set. I think it'd be intuitive because e.g. when you scroll a list probably you don't want the gesture to be activated.

File: src/lv_core/lv_obj.h
Comment (embeddedt) [inline src/lv_core/lv_obj.h]: Typo here and with `gusture_parent`.

**Review by embeddedt — CHANGES_REQUESTED** [review]

**Comment by Simen-z** [comment]
@kisvegabor Yes,gestures can be used for screen-sized (the touch screen needs to support multiple contacts), or for cutting the screen.
When you set drag_in_prog, you can clear gesture_in_prog, you're right.

File: src/lv_core/lv_obj.h
Comment (Simen-z) [reply src/lv_core/lv_obj.h]: Has been modified.

Comment (Simen-z) [reply src/lv_core/lv_obj.h]: Do I need to resend the pull request?

Comment (kisvegabor) [reply src/lv_core/lv_obj.h]: > Do I need to resend the pull request?

No, it's fine to push commits here.

**Comment by kisvegabor** [comment]
> the touch screen needs to support multiple contacts

Why is it required?

**Comment by Simen-z** [comment]
oh,sorry,This is a problem with the translation software, which should be called Multi-Touch Gestures.

**Comment by kisvegabor** [comment]
I see, but why do we need multiple touch points here?

**Comment by kisvegabor** [comment]
I've changed a few things:
1. Send gesture immediately when the limit is reached (not on release)
2. Set `gesture_parent = 1` by default
3. Remove `lv_obj_set/get_gesture`. User can ignore gestures by not reacting on the events
4. Not send gestures if `drag_in_prog`
5. Save the related defines in the driver to allow runtime configuration

Let me know what you think.

**Comment by Simen-z** [comment]
1.I think not making release judgment may cause some problems. For example, when cutting the screen by gesture, the screen has switched to the second screen by gesture event, which means the finger has not left the screen, which may cause the wrong touch to the second screen.If there is a way to avoid such a situation, it(not on release) is ok to do so.
2.3.4.5.No problem.

**Comment by Simen-z** [comment]
**I see, but why do we need multiple touch points here?**
Don't you mean using two fingers(touch points) to scale the screen?If that's not what it means, it doesn't need to be multiple touch points.

File: src/lv_core/lv_obj.h
Comment (kisvegabor) [reply src/lv_core/lv_obj.h]: I've fixed "gusuture" typos.

**Comment by kisvegabor** [comment]
> I think not making release judgment may cause some problems. For example, when cutting the screen by gesture, the screen has switched to the second screen by gesture event, which means the finger has not left the screen, which may cause the wrong touch to the second screen.If there is a way to avoid such a situation, it(not on release) is ok to do so.

That's correct. Fortunately, there is a built-in method to solve it. See 79e342c

> Don't you mean using two fingers(touch points) to scale the screen?

Two (or more ) finger handling is conceptually missing from the library now so we should be fine with one finger gestures for now.

**Comment by TridentTD** [comment]
I have just tried by [79e342c](https://github.com/littlevgl/lvgl/commit/79e342c807b68a83fc08dc72e51e226af3314b44).
for draging menu from the top edge.

If `proc->wait_until_release = 1;` in `indev_gesture()`,
after LV_EVENT_GESTURE occurs,  LV_EVENT_PRESSING or other events is disappeared.

Wish `proc->wait_until_release = 0;` in `indev_gesture()`,
is it suitable for still can receive LV_EVENT_PRESSING ?

**Comment by kisvegabor** [comment]
> Wish proc->wait_until_release = 0; in indev_gesture(),
is it suitable for still can receive LV_EVENT_PRESSING ?

Actually you can call `lv_indev_wait_release(indev)` in `LV_EVENT_GESTURE` if you need  it so we can remove `proc->wait_until_release = 1`.

EDIT:
I've just removed it.

**Comment by embeddedt** [comment]
Is this ready for me to fix conflicts on and merge?

**Comment by barbiani** [comment]
Hi,

I tested it and it did not feel correct when the top-down drag is aborted.
At the slighest up move the panel disappears instead of switching the move
direction.

**Review by kisvegabor — APPROVED** [review]
Ready from my side!

**Comment by embeddedt** [comment]
Great! I will merge it as soon as I get a chance.

File: src/lv_core/lv_indev.c
Comment (ckielstra) [inline src/lv_core/lv_indev.c]:
```suggestion
	/*If gesture parent is active check recursively the drag_parent attribute*/
```

File: src/lv_core/lv_obj.h
Comment (ckielstra) [inline src/lv_core/lv_obj.h]:
```suggestion
* Enable to use parent for gesture related operations.
```

File: src/lv_core/lv_obj.h
Comment (ckielstra) [inline src/lv_core/lv_obj.h]:
```suggestion
* @param en true: enable the 'gesture parent' for the object
```

File: src/lv_core/lv_obj.h
Comment (ckielstra) [inline src/lv_core/lv_obj.h]:
```suggestion
* If trying to gesture the object the parent will be moved instead
```

File: src/lv_core/lv_obj.h
Comment (ckielstra) [inline src/lv_core/lv_obj.h]:
```suggestion
void lv_obj_set_gesture_parent(lv_obj_t * obj, bool en);
```

File: src/lv_draw/lv_draw_arc.c
Comment (ckielstra) [inline src/lv_draw/lv_draw_arc.c]: Actually a boolean would be big enough. :-)
Suggest to use an uint8_t instead.

File: src/lv_core/lv_indev.c
Comment (ckielstra) [inline src/lv_core/lv_indev.c]:
```suggestion
```

**Comment by embeddedt** [comment]
Thanks for the review, @ckielstra! I agree with all of the suggestions and will hold off on merging for now.

**Comment by kisvegabor** [comment]
@ckielstra Thank you for the suggestions.
@LanFengXiMenChuiXue, for the sake of simplicity,  can you merge them? (only you have right to do it.)

**Review by Simen-z — COMMENTED** [review]
review

**Comment by Simen-z** [comment]
@kisvegabor I can't merge them.Can you help me?
Github shows that I do not have write permissions:
捕获
Github warning:
This branch has conflicts that must be resolved
Only those with write access to this repository can merge pull requests.

**Comment by embeddedt** [comment]
That button is for us to merge the pull request into our repository. Your commit has already been applied to the pull request (it happens automatically when you push to the branch).

File: src/lv_draw/lv_draw_arc.c
Comment (embeddedt) [reply src/lv_draw/lv_draw_arc.c]: I agree; `thick_corr` should be a `uint8_t`.

**Comment by embeddedt** [comment]
Looks pretty good! I just left one comment (it seems you missed the original one by @ckielstra).

**Comment by Simen-z** [comment]
So what do I need to do now?

**Comment by embeddedt** [comment]
Please apply [this change](https://github.com/littlevgl/lvgl/pull/1300/files#r363738598) and then I can merge the PR into `dev-7.0`.

**Comment by Simen-z** [comment]
Ok. Is there anything else I can do?

**Comment by Simen-z** [comment]
I'm going home, because there will be no subway soon.@_@. If there is anything else that needs to be changed, please tell me, I will change tomorrow.Thank you.

**Review by embeddedt — APPROVED** [review]

**Comment by embeddedt** [comment]
Looks good! I will merge the changes soon.

Thank you for your contribution! :slightly_smiling_face:

**Comment by embeddedt** [comment]
Merged in https://github.com/littlevgl/lvgl/commit/400ca85c67d74a2807ff76571e2bc1c444e47f82.

**Pull request closed without merging**
